drivers/input/ajoystick.c

```diff
@@ -526,12 +461,7 @@ static ssize_t ajoy_read(FAR struct file *filep, FAR char *buffer,
 
   /* Get exclusive access to the driver structure */
 
-  ret = ajoy_takesem(&priv->au_exclsem);
-  if (ret < 0)
-    {
-      ierr("ERROR: ajoy_takesem failed: %d\n", ret);
-      return ret;
-    }
+  flags = enter_critical_section();
```

Comment:
Is `lower->al_sample(lower, (FAR struct ajoy_sample_s *)buffer);` -> `al_sample` implementation ready that now it is called from inside the critical section?

Reply:
Yes, al_sample is already called inside critical section before ajoy_sample:
https://github.com/apache/incubator-nuttx/pull/5204/files#diff-5803ad0a8a93175597d14948ffe9a46f4e1ad28b59d28bcb568f56553b1fa8f7R249-R256

Reply:
Sorry I do not see that `al_sample` is called from `ajoy_sample`

Reply:
I agree with buttons and djoystick changes, but first look here leads me to the next findings in `boards/arm/stm32/nucleo-f4x1re/src/stm32_ajoystick.c`:
```
static int ajoy_sample(FAR const struct ajoy_lowerhalf_s *lower,
                       FAR struct ajoy_sample_s *sample)
{
#ifndef NO_JOYSTICK_ADC
  struct adc_msg_s adcmsg[MAX_ADC_CHANNELS];
  FAR struct adc_msg_s *ptr;
  ssize_t nread;
  ssize_t offset;
  int have;
  int i;

  /* Read all of the available samples (handling the case where additional
   * channels are enabled).
   */

  nread = file_read(&g_adcfile, adcmsg,
                    MAX_ADC_CHANNELS * sizeof(struct adc_msg_s));
  if (nread < 0)
    {
      if (nread != -EINTR)
        {
          ierr("ERROR: read failed: %d\n", (int)nread);
        }

      return nread;
    }
...
```
So the question is: Is it safe to call `file_read` from the critical section?

Reply:
It's fine to call file_read inside critical section, since the implementation of critical section in nuttx allow the thread enter the wait state while the critical section is hold.

Reply:
It will work only because we do not have any manipulations with `priv` after `al_sample` call. Otherwise in theory if we enter a critical section and then pend inside the `file_read` call and during that time `ajoy_close` is called, then after returning from `file_read` the `priv` would point to already free-ed memory.
But this is not a theoretical case, so I'm fine with this change even if I do not like cases like this.

Reply:
vfs layer can't handle this case correctly in the complex case. To avoid the used after free, the better solution is:

1. Add a reference count to struct file
2. Initialize the reference count to 1 in file_open
3. Decrease the reference count in file_close
4. Increase count before call file_xxx and decrease after
5. Release the resource only when the reference count become zero

This way, the real release happen(delayed) after file_read return when file_close is called during file_read is executing.

Reply:
Yes. The real use case would require reference counter. Or we could still keep a semaphore for `priv` protection and critical section to protect `opriv` manipulations.
I'm just thinking what if `ajoy_interrupt` fires while `al_sample` is pending on `file_read` call? I briefly looked thru the code and see that seems like nothing harmful, but I still have some doubts.

Please add more reviewers to this change before merging it.

Reply:
> Yes. The real use case would require reference counter. Or we could still keep a semaphore for `priv` protection and critical section to protect `opriv` manipulations.

So, which resource is protected by semaphore? The problem this PR want to fix is that the original code protect the same resource some time with semaphore, other time with the critical section, for example:

1. ajoy_open/ajoy_close hold au_exclsem before modifying ao_flink
2. ajoy_interrupt(ajoy_sample) hold the critical section before access ao_flink

so, the final result is that no protection at all. You can't protect the same resource by two semaphores or mi of semaphore and critical section. Since we need allow ajoy_interrupt is called from the interrupt context directly, the critical section is the only choice, and then semaphore must be removed to avoid the potential misuse like the above in the future. Actually, I always prefer to use semaphore to protect resource, but has to keep the critical section here due to the requirement of ajoy_interrupt.

> I'm just thinking what if `ajoy_interrupt` fires while `al_sample` is pending on `file_read` call? I briefly looked thru the code and see that seems like nothing harmful, but I still have some doubts.
> 

It's fine in this case. What can't handle is that someone call close while other file api is executing. This problem isn't specific to joystick/button drivers, it's a common issue to the whole file system, and can't handle inside the driver level. The old code try to fix this problem with ao_closing, but it impossible to work.

> Please add more reviewers to this change before merging it.

Reply:
IMO the semaphore protects `priv` and critical section protects `opriv` + iteration thru the linked list

Reply:
The link list is just a demo, if you evaluate each field on by one, you will find that all most fields touch in both the thread context and interrupt context.

Reply:
Probably you are right, but the usage of semaphore gives us exclusive access with waiting for a task context while critical section resolves mutual exclusion between task and interrupt context inside the driver. I mean that only one of multiple callers of `ajoy_read` will call `lower->al_sample()`+return from `al_sample` with semaphore and multiple callers will call `lower->al_sample()` in case if critical section is used and `al_sample()` call a pending primitive. That is conceptually difference that I'm trying to clarify the behavior.

Reply:
@patacongo I will appreciate your input here.

Reply:
The current PR does not reflect the actual code and should be rebased, so I will write a comment taking into assumption that PR is based on the latest mainline.

I'm not sure if what I describe is a valid use case, so please correct me if I'm wrong.

On the latest mainline the next code is present:
```
  lower = priv->au_lower;
  DEBUGASSERT(lower && lower->al_sample);
  ret = lower->al_sample(lower, (FAR struct ajoy_sample_s *)buffer);
  if (ret >= 0)
    {
      opriv->ao_pollpending = false;
      ret = sizeof(struct ajoy_sample_s);
    }
```
So I'm thinking if that can be a situation when we enter a critical section and then `al_sample` pends. While we are waiting in `al_sample` the `ajoy_close` is called and that leads to `kmm_free(opriv);` so after we return from `al_sample` the `opriv->ao_pollpending = false;` can access released memory.

In general I do not understand what we protect here with a critical section? Maybe critical section can be removed at all from `ajoy_read`?

Reply:
al_sample is impossible to pend since it may be called inside interrupt handler. This is the root reason why we must protect it with critical section, not semaphore.

Reply:
`lower->al_sample` is not called inside interrupt handler. At least what I see from the code now.

Reply:
ajoy_interrupt will be called inside interrupt handler which call lower->al_sample through ajoy_sample indirectly. If ajoy_interrupt is called only in thread context. All protection can be switch to the semaphore instead.

Reply:
Sorry, but `ajoy_interrupt` calls `ajoy_sample`, but not `lower->al_sample`. The `lower->al_sample` is called only from `ajoy_read`. So interrupt handles only a change of buttons and wake-up reader and then reader reads the device data.

Reply:
Oh, yes, but ajoy_sample call lower->al_buttons:
https://github.com/apache/incubator-nuttx/pull/5204/files#diff-5803ad0a8a93175597d14948ffe9a46f4e1ad28b59d28bcb568f56553b1fa8f7R241
I would expect that al_sample has the similar implementation requirement as al_buttons. Do you think the assumption reasonable?

Reply:
No. And that is why I'm keeping this discussion open. `al_buttons` access only board GPIOs via arch specific calls and do not use pending primitives (example sama5/giant-board: `ajoy_buttons` calls `sam_pioread`), while `al_sample` can be implemented as a reader from a file device and can pend (example sama5/giant-board: `ajoy_sample` calls `file_read` and `ajoy_buttons` at the end).

Reply:
Ok, I move enter_critical_section after call low->al_sample.

Reply:
The situation that I described is still not handled. We need to check if driver is still opened after `lower->al_sample(lower, (FAR struct ajoy_sample_s *)buffer);` returns and if we can execute `opriv->ao_pollpending = false;`.

Reply:
In this case, opriv is already freed by ajoy_close, how to detect it with the freed memory? All file handles(file system, char driver, block driver, socket) has the same issue, as I said before it need handle in the general way.

Reply:
We can handle this by using a semaphore to protect `priv` and critical section to protect `opriv` list.

Could you please write more about "All file handles(file system, char driver, block driver, socket) has the same issue" (some examples) so I can take a look?

Reply:
You can't fix the race condition in all case. How to handle this sequence?

1. Thread A just enter ajoy_close and switch to thread B
2. Thread B just enter ajoy_read and switch to thread A again
3. Thread A continue ajoy_close and release opriv
4. Thread B resume and touch the freed memory

Basically, the resource release is very very special, you can't protect the resource by the normal approach.

Reply:
So what is not handled in this case?
1. Thread A just enter ajoy_close and switch to thread B. -- I assume that this is done either before `ajoy_takesem(&priv->au_exclsem);` or inside the `ajoy_takesem(&priv->au_exclsem);` call.
2. Thread B just enter ajoy_read and switch to thread A again. -- This will either switch before `ret = ajoy_takesem(&priv->au_exclsem);` or inside the `ret = ajoy_takesem(&priv->au_exclsem);` call.
3. Thread A continue ajoy_close and release opriv -- Yes, so in this case we need to access `filep->f_priv` (`opriv`) after `ret = ajoy_takesem(&priv->au_exclsem);` is successful and `filep->f_priv` should be set to `NULL` after memory is freed.
4. Thread B resume and touch the freed memory -- this will not happen if `ajoy_takesem(&priv->au_exclsem);` is successfully taken and `filep->f_priv` is accessed after that.

My assumption is next: The `priv->au_exclsem` protects any modification of `node->i_private` data and `filep->f_priv` data. If there is no modification, then we do not need to take `priv->au_exclsem`. We can try to leverage critical section approach.
In other words the `ajoy_read` should look like:
```
static ssize_t ajoy_read(FAR struct file *filep, FAR char *buffer,
                         size_t len)
{
  FAR struct inode *inode;
  FAR struct ajoy_open_s *opriv;
  FAR struct ajoy_upperhalf_s *priv;
  FAR const struct ajoy_lowerhalf_s *lower;
  irqstate_t flags;
  int ret;

  inode = filep->f_inode;
  DEBUGASSERT(inode->i_private);
  priv  = (FAR struct ajoy_upperhalf_s *)inode->i_private;

  /* Make sure that the buffer is sufficiently large to hold at least one
   * complete sample.
   *
   * REVISIT:  Should also check buffer alignment.
   */

  if (len < sizeof(struct ajoy_sample_s))
    {
      ierr("ERROR: buffer too small: %lu\n", (unsigned long)len);
      return -EINVAL;
    }

  /* Read and return the current state of the joystick buttons */

  lower = priv->au_lower;
  DEBUGASSERT(lower && lower->al_sample);
  ret = lower->al_sample(lower, (FAR struct ajoy_sample_s *)buffer);

  /* Get exclusive access to the driver structure */

  if (ret >= 0)
    {
      flags = enter_critical_section();

      DEBUGASSERT(filep && filep->f_inode);
      opriv = filep->f_priv;

      if (opriv != NULL)
        {
          opriv->ao_pollpending = false;
          ret = sizeof(struct ajoy_sample_s);
        }
      else
        {
          ret = -EPIPE;
        }

      leave_critical_section(flags);
    }

  return (ssize_t)ret;
}
```

Reply:
And `ajoy_close` definitely be modified to have
```
kmm_free(opriv);
filep->f_priv = NULL;
```

Reply:
> So what is not handled in this case?
> 
> 1. Thread A just enter ajoy_close and switch to thread B. -- I assume that this is done either before `ajoy_takesem(&priv->au_exclsem);` or inside the `ajoy_takesem(&priv->au_exclsem);` call.

I mean before take au_exclsem.

> 2. Thread B just enter ajoy_read and switch to thread A again. -- This will either switch before `ret = ajoy_takesem(&priv->au_exclsem);` or inside the `ret = ajoy_takesem(&priv->au_exclsem);` call.

I mean before take au_exclsem too.

> 3. Thread A continue ajoy_close and release opriv -- Yes, so in this case we need to access `filep->f_priv` (`opriv`) after `ret = ajoy_takesem(&priv->au_exclsem);` is successful and `filep->f_priv` should be set to `NULL` after memory is freed.

Yes.

> 4. Thread B resume and touch the freed memory -- this will not happen if `ajoy_takesem(&priv->au_exclsem);` is successfully taken and `filep->f_priv` is accessed after that.
> 

Once ajoy_close return at the step 3, filep may be reused by VFS layer and then f_priv may point to a different struct, which make the check of filep->f_priv useless.

> My assumption is next: The `priv->au_exclsem` protects any modification of `node->i_private` data and `filep->f_priv` data. If there is no modification, then we do not need to take `priv->au_exclsem`. We can try to leverage critical section approach. In other words the `ajoy_read` should look like:
> 
> ```
>   /* Get exclusive access to the driver structure */
> 
>   if (ret >= 0)
>     {
>       flags = enter_critical_section();
> 
>       DEBUGASSERT(filep && filep->f_inode);
>       opriv = filep->f_priv;
> 
>       if (opriv != NULL)
>         {
>           opriv->ao_pollpending = false;
>           ret = sizeof(struct ajoy_sample_s);
>         }
>       else
>         {
>           ret = -EPIPE;
>         }
> 
>       leave_critical_section(flags);
>     }
> 
>   return (ssize_t)ret;
> }
> ```

neither semaphore nor critical section can resolve the above sequency.

Reply:
@xiaoxiang781216 thank you. Yes, the reuse of data by VFS is still an unhanded problem. Now I agree to keep your previous version
```
static ssize_t ajoy_read(FAR struct file *filep, FAR char *buffer,
                         size_t len)
{
  FAR struct inode *inode;
  FAR struct ajoy_open_s *opriv;
  FAR struct ajoy_upperhalf_s *priv;
  FAR const struct ajoy_lowerhalf_s *lower;
  int ret;

  DEBUGASSERT(filep && filep->f_inode);
  opriv = filep->f_priv;
  inode = filep->f_inode;
  DEBUGASSERT(inode->i_private);
  priv  = (FAR struct ajoy_upperhalf_s *)inode->i_private;

  /* Make sure that the buffer is sufficiently large to hold at least one
   * complete sample.
   *
   * REVISIT:  Should also check buffer alignment.
   */

  if (len < sizeof(struct ajoy_sample_s))
    {
      ierr("ERROR: buffer too small: %lu\n", (unsigned long)len);
      return -EINVAL;
    }

  /* Get exclusive access to the driver structure */

  flags = enter_critical_section();

  /* Read and return the current state of the joystick buttons */

  lower = priv->au_lower;
  DEBUGASSERT(lower && lower->al_sample);
  ret = lower->al_sample(lower, (FAR struct ajoy_sample_s *)buffer);
  if (ret >= 0)
    {
      opriv->ao_pollpending = false;
      ret = sizeof(struct ajoy_sample_s);
    }

  leave_critical_section(flags);
  return (ssize_t)ret;
}
```
and resolve a limitation of a driver based on VFS device in a different way. This will be a limitation for now.

Reply:
No problem, I enjoy very much this kind of technical conversation.

Reply:
Done.